### PR TITLE
More string API updates

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR updates Communicator.h, Initialize.h, ImplicitContext.h and Plugin.h.

### CODE REVIEW

File: slice/Ice/Context.ice [hunk 1]
```diff
@@ -21,8 +19,8 @@
 
 module Ice
 {
-    /// A request context. <code>Context</code> is used to transmit metadata about a request from the server to the
```
Comment (bernardnormier) [inline slice/Ice/Context.ice:24]: The old description was bogus.

File: slice/Ice/Context.ice [hunk 2]
```diff
@@ -21,8 +19,8 @@
...
-    /// client, such as Quality-of-Service (QoS) parameters. Each operation on the client has a <code>Context</code> as
-    /// its implicit final parameter.
+    /// A string-string dictionary, used to transmit additional information with an Ice request. This context is only
+    /// transmitted with requests, from clients to servers; the Ice protocol does not provide response contexts.
+    ["cpp:type:std::map<std::string, std::string, std::less<>>"]
```
Comment (bernardnormier) [inline slice/Ice/Context.ice:24]: The addition of `std::less<>` is the most disruptive part of this update.

**Review by pepone — APPROVED** [review]

File: cpp/src/Ice/PluginManagerI.cpp [hunk 3]
```diff
@@ -19,43 +19,43 @@ const char* const Ice::PluginManagerI::_kindOfObject = "plugin";
...
         factories->insert(make_pair(name, factory));
         if (loadOnInit)
         {
-            if (loadOnInitialization == 0)
+            if (!loadOnInitialization)
             {
                 loadOnInitialization = new vector<string>();
             }
-            loadOnInitialization->push_back(name);
+            loadOnInitialization->push_back(std::move(name));
```
Comment (externl) [inline cpp/src/Ice/PluginManagerI.cpp:58]: Could use emplace_back here.

Comment (bernardnormier) [reply cpp/src/Ice/PluginManagerI.cpp:58]: push_back is simpler.

File: cpp/src/Ice/PluginManagerI.cpp [hunk 4]
```diff
@@ -400,14 +400,14 @@ Ice::PluginManagerI::loadPlugin(const string& name, const string& pluginSpec, St
 }
 
 Ice::PluginPtr
-Ice::PluginManagerI::findPlugin(const string& name) const
+Ice::PluginManagerI::findPlugin(string_view name) const
 {
-    for (PluginInfoList::const_iterator p = _plugins.begin(); p != _plugins.end(); ++p)
+    for (auto p : _plugins)
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/Ice/PluginManagerI.cpp:405]: Fixed as suggested.

File: cpp/src/Ice/Reference.cpp [hunk 5]
```diff
@@ -1125,6 +1124,8 @@ IceInternal::RoutableReference::toProperty(const string& prefix) const
         }
     }
 
+    properties[std::move(prefix)] = toString();
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/Ice/Reference.cpp:1127]: Updated as suggested.

File: cpp/src/Ice/PluginManagerI.cpp [hunk 6]
```diff
@@ -19,43 +19,43 @@ const char* const Ice::PluginManagerI::_kindOfObject = "plugin";
 
 namespace
 {
-    map<string, PluginFactory>* factories = 0;
-    vector<string>* loadOnInitialization = 0;
+    map<string, PluginFactory>* factories = nullptr;
+    vector<string>* loadOnInitialization = nullptr;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/Ice/PluginManagerI.cpp:23]: That's not what this code is about. It's about static/lazy initialization and cleanup during static destruction.

**Review by externl — APPROVED** [review]

File: cpp/src/Ice/PluginManagerI.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (pepone) [reply cpp/src/Ice/PluginManagerI.cpp:23]: We can still do lazy initialization using unique_ptr, which would be destroyed during static destruction.

**Pull request merged**